_posts/2019-02-05-configurable-builds-part-1.md

```diff
@@ -0,0 +1,284 @@
+While there are many kinds of "ways" you might want to build your project, in
+practice most developers want the flexibility to target different devices,
+OSes, CPUs, and other machine properties. The general term for this is
+*platforms*. Being able to reason about how your code interacts with platforms
+is the basis of [multi-platform software](https://en.wikipedia.org/wiki/Cross-platform_software).
+
+*Toolchains* are the set of programs that build your code. A C++ toolchain
+includes a compiler and linker and the flags that invoke them "correctly".
+Different platforms require different toolchains. gcc might be the compiler of
+choice for Linux while Xcode is used on the Mac. Even the same platform might
+use different toolchains. For example, maybe you want to use an experimental
+compiler.
+
+Not only are these concepts cross-language, but it's *important* that languages
+treat them consistently. Otherwise you get an uncoordinated mess of
+language-specific code that makes it hard to combine different languages in a
+single project.
+
+One of the reasons
+
+    $ bazel build //my:binary --cpu=arm --crosstool_top=//my:custom_toolchain --define MYFEATURE=1
+
+doesn't work well is that `--cpu` and `--crosstool_top` are C++-specific. What
+if you want to build Java libraries with C++ dependencies for both Linux and
+Android? This impacts not just C++, but also the JDK and maybe even which
+support libraries you need.
+
+Not only are `--cpu` and `-crosstool_top` insufficiently expressive, but they
+might actively sabotoge you. Consider the following use of
+[select](https://docs.bazel.build/versions/master/configurable-attributes.html):
+
+```python
+java_library(
+    name = "supporting_library",
+	deps = select({
+	    ":android": [":extra_required_android_deps"],
+		"//conditions:default": []
+    }))
+```
+
+If you use `--cpu` and `crosstool_top` to define what `:android` means, what
+happens when an app supports a new Android phone with a different CPU?
+`:android` won't trigger because `--cpu` no longer matches, the app won't get
+required Android support libraries, and it will crash.
+
+### A Better Way
+
+The configurability's team's platform work, led by [@katre](https://github.com/katre),
+lets you declare exactly what you want in a way all rules understand. It
+doesn't even matter if no one's ever heard of your platform.
+
+With the new API, our example can be rewritten as:
+
+```python
+$ cat platforms/BUILD
+constraint_setting(name = "os")
+constraint_value(name = "android", constraint_setting = ":os")
+constraint_value(name = "linux", constraint_setting = ":os")
+
+constraint_setting(name = "device")
+constraint_value(name = "phone", constraint_setting = ":device")
+constraint_value(name = "tablet", constraint_setting = ":device")
+
+platform(name = "pixel3", constraint_values = [":android", ":phone"])
+```
+
+```python
+$ cat helpers/BUILD
+config_setting(name = "android", constraint_values = "//platforms:android")
+java_library(
```

Comment:
Indentation's weird here too. Should ideally be

```
    deps = select({
        ":android": [":extra_required_android_deps"],
        "//conditions:default": [],
    }),
)
```

Reply:
Text editor auto-tabbing. :(